contracts/vaults/TreasuryVault/RibbonAutocallVault.sol

```diff
@@ -0,0 +1,461 @@
+
+    event OptionTypeSet(OptionType optionType);
+
+    event CouponStateSet(
+        CouponType couponType,
+        uint256 newAutocallBarrierPCT,
+        uint256 newCouponBarrierPCT
+    );
+
+    event PeriodSet(uint256 period, uint256 newPeriod);
+
+    event ObservationPeriodFreqSet(
+        uint256 observationPeriodFreq,
+        uint256 newObservationPeriodFreq
+    );
+
+    /************************************************
+     *  CONSTRUCTOR & INITIALIZATION
+     ***********************************************/
+
+    /**
+     * @notice Initializes the contract with immutable variables
+     * @param _usdc is the USDC contract
+     * @param _oTokenFactory is the contract address for minting new opyn option types (strikes, asset, expiry)
+     * @param _gammaController is the contract address for opyn actions
+     * @param _marginPool is the contract address for providing collateral to opyn
+     */
+    constructor(
+        address _usdc,
+        address _oTokenFactory,
+        address _gammaController,
+        address _marginPool
+    )
+        RibbonTreasuryVaultLite(
+            _usdc,
+            _oTokenFactory,
+            _gammaController,
+            _marginPool
+        )
+    {
+        ORACLE = IOracle(IController(_gammaController).oracle());
+    }
+
+    /**
+     * @notice Initializes the OptionVault contract with storage variables.
+     * @param _initParams is the struct with vault initialization parameters
+     * @param _vaultParams is the struct with vault general data
+     * @param _optionType is type of the next put option
+     * @param _couponState is the coupon state
+     */
+    function initialize(
+        VaultLifecycleTreasury.InitParams calldata _initParams,
+        Vault.VaultParams calldata _vaultParams,
+        OptionType _optionType,
+        CouponState calldata _couponState,
+        uint256 _observationPeriodFreq,
+        address _autocallSeller
+    ) external initializer {
+        _initialize(_initParams, _vaultParams);
+        _verifyCouponState(
+            _couponState.nextCouponType,
+            _couponState.nextAutocallBarrierPCT,
+            _couponState.nextCouponBarrierPCT
+        );
+
+        require(_autocallSeller != address(0), "!_autocallSeller");
+        require(
+            _observationPeriodFreq > 0 &&
+                _observationPeriodFreq % 1 days == 0 &&
+                _observationPeriodFreq <= period,
+            "!_observationPeriodFreq"
+        );
+
+        putOption.nextOptionType = _optionType;
+        couponState.nextCouponType = _couponState.nextCouponType;
+        couponState.nextAutocallBarrierPCT = _couponState.autocallBarrierPCT;
+        couponState.nextCouponBarrierPCT = _couponState.couponBarrierPCT;
+
+        nextObservationPeriodFreq = _observationPeriodFreq;
+        autocallSeller = _autocallSeller;
+        numTotalObservationPeriods = period / _observationPeriodFreq;
```

Comment:
similar to [this](https://github.com/ribbon-finance/ribbon-v2/pull/311/files#r1181614904)